File: mekaneck/virtualbrain/engine/state_manager.py
"""
StateManager: State transitions and validation.

Manages mental state transitions, validates states, and maintains
state history for the Virtual Brain.
"""

from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any
import numpy as np

from ..core.types import MentalState, SCoord, PartitionCoord
from ..operators.partition_ops import partition_to_sentropy, D_CAT


@dataclass
class TransitionResult:
    """Result of a state transition."""

    success: bool
    from_state: MentalState
    to_state: MentalState
    distance: float
    metadata: Dict[str, Any] = field(default_factory=dict)


class StateManager:
    """
    Manages mental state transitions and validation.

    Ensures state consistency, tracks history, and provides
    utilities for state manipulation.
    """

    def __init__(self, max_history: int = 1000):
        """
        Initialize state manager.

        Args:
            max_history: Maximum states to keep in history
        """
        self.max_history = max_history
        self.history: List[MentalState] = []
        self.current_state: Optional[MentalState] = None

    def initialize(
        self,
        s_coord: Optional[SCoord] = None,
        partition: Optional[PartitionCoord] = None,
    ) -> MentalState:
        """
        Initialize with a new state.

        Args:
            s_coord: Initial S-entropy coordinate
            partition: Initial partition coordinate

        Returns:
            Initial MentalState
        """
        if s_coord is None:
            s_coord = SCoord(sk=0.5, st=0.5, se=0.5)

        state = MentalState.initial(s_coord=s_coord, partition=partition)
        self.current_state = state
        self.history = [state]

        return state

    def transition(
        self,
        target_s_coord: Optional[SCoord] = None,
        target_partition: Optional[PartitionCoord] = None,
        new_gamma: Optional[float] = None,
        new_gamma_f: Optional[float] = None,
    ) -> TransitionResult:
        """
        Transition to new state.

        Args:
            target_s_coord: Target S-entropy coordinate
            target_partition: Target partition coordinate
            new_gamma: New phase coherence
            new_gamma_f: New frequency coherence

        Returns:
            TransitionResult with transition details
        """
        if self.current_state is None:
            raise RuntimeError("State manager not initialized")

        old_state = self.current_state

        # Compute transition distance
        distance = 0.0
        if target_s_coord and old_state.s_coord:
            distance = old_state.s_coord.distance(target_s_coord)
        elif target_partition and old_state.partition:
            distance = D_CAT(old_state.partition, target_partition)

        # Create new state
        new_state = old_state.transition_to(
            new_s_coord=target_s_coord,
            new_partition=target_partition,
            new_gamma=new_gamma,
            new_gamma_f=new_gamma_f,
        )

        # Update current and history
        self.current_state = new_state
        self._add_to_history(new_state)

        return TransitionResult(
            success=True,
            from_state=old_state,
            to_state=new_state,
            distance=distance,
        )

    def validate_state(self, state: MentalState) -> bool:
        """
        Validate that state is consistent.

        Checks:
        - gamma in [0, 1]
        - gamma_f in [0, 1]
        - p_decay in [0, 1]
        - t_decay in [0, 1]
        - S-coordinates in [0, 1]^3

        Args:
            state: State to validate

        Returns:
            True if state is valid
        """
        # Check coherence values
        if not (0 <= state.gamma <= 1):
            return False
        if not (0 <= state.gamma_f <= 1):
            return False
        if not (0 <= state.p_decay <= 1):
            return False
        if not (0 <= state.t_decay <= 1):
            return False

        # Check S-coordinates
        if state.s_coord:
            if not (0 <= state.s_coord.sk <= 1):
                return False
            if not (0 <= state.s_coord.st <= 1):
                return False
            if not (0 <= state.s_coord.se <= 1):
                return False

        return True

    def get_trajectory(self) -> List[SCoord]:
        """
        Get S-entropy trajectory from history.

        Returns:
            List of S-coordinates from history
        """
        return [s.s_coord for s in self.history if s.s_coord is not None]

    def get_consciousness_history(self) -> np.ndarray:
        """
        Get consciousness values from history.

        Returns:
            Array of consciousness values
        """
        return np.array([s.consciousness for s in self.history])

    def get_memory_history(self) -> np.ndarray:
        """
        Get memory values from history.

        Returns:
            Array of memory values
        """
        return np.array([s.M for s in self.history])

    def find_state_at_time(self, t: float) -> Optional[MentalState]:
        """
        Find state closest to given time.

        Args:
            t: Target time

        Returns:
            MentalState closest to time t, or None
        """
        if not self.history:
            return None

        closest = min(self.history, key=lambda s: abs(s.timestamp - t))
        return closest

    def compute_memory_differential(self) -> float:
        """
        Compute current rate of memory change.

        Returns:
            dM/dt at current state
        """
        if len(self.history) < 2:
            return 0.0

        recent = self.history[-1]
        previous = self.history[-2]

        dt = recent.timestamp - previous.timestamp
        if dt == 0:
            return 0.0

        return (recent.M - previous.M) / dt

    def enter_dream_mode(self) -> MentalState:
        """
        Transition current state to dream mode.

        Returns:
            Dream state
        """
        if self.current_state is None:
            raise RuntimeError("State manager not initialized")

        dream_state = self.current_state.enter_dream()
        self.current_state = dream_state
        self._add_to_history(dream_state)

        return dream_state

    def wake_up(self, perception_level: float = 1.0) -> MentalState:
        """
        Transition from dream to awake.

        Args:
            perception_level: Initial perception level

        Returns:
            Awake state
        """
        if self.current_state is None:
            raise RuntimeError("State manager not initialized")

        wake_state = self.current_state.wake(perception_level)
        self.current_state = wake_state
        self._add_to_history(wake_state)

        return wake_state

    def _add_to_history(self, state: MentalState):
        """Add state to history, maintaining max size."""
        self.history.append(state)
        if len(self.history) > self.max_history:
            self.history = self.history[-self.max_history :]

    def clear_history(self):
        """Clear state history."""
        if self.current_state:
            self.history = [self.current_state]
        else:
            self.history = []

    def summary(self) -> Dict[str, Any]:
        """
        Get summary of current state and history.

        Returns:
            Dictionary with state summary
        """
        if self.current_state is None:
            return {"initialized": False}

        return {
            "initialized": True,
            "current_consciousness": self.current_state.consciousness,
            "current_memory": self.current_state.M,
            "is_dreaming": self.current_state.is_dreaming,
            "is_awake": self.current_state.is_awake,
            "history_length": len(self.history),
            "total_time": self.current_state.timestamp,
        }

File: virtual_brain/crates/vb-engine/src/state_manager.rs
//! State Manager: Tracks state transitions and history.

use serde::{Deserialize, Serialize};
use vb_core::types::{MentalState, PartitionCoord, SCoord};

/// Result of a state transition.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransitionResult {
    /// Whether transition succeeded
    pub success: bool,
    /// State before transition
    pub from_state: Option<SCoord>,
    /// State after transition
    pub to_state: SCoord,
    /// Distance traveled
    pub distance: f64,
    /// Additional metadata
    pub metadata: std::collections::HashMap<String, String>,
}

/// Manages state transitions and history.
#[derive(Debug, Clone)]
pub struct StateManager {
    /// Current mental state
    current_state: MentalState,
    /// History of states
    history: Vec<MentalState>,
    /// Maximum history size
    max_history: usize,
}

impl StateManager {
    /// Create new state manager.
    pub fn new(max_history: usize) -> Self {
        Self {
            current_state: MentalState::default(),
            history: Vec::new(),
            max_history,
        }
    }

    /// Initialize with a specific state.
    pub fn initialize(&mut self, s_coord: Option<SCoord>, partition: Option<PartitionCoord>) {
        self.current_state = MentalState::initial(s_coord, partition);
        self.history.clear();
        self.history.push(self.current_state.clone());
    }

    /// Get current state.
    pub fn current(&self) -> &MentalState {
        &self.current_state
    }

    /// Get state history.
    pub fn history(&self) -> &[MentalState] {
        &self.history
    }

    /// Transition to new S-coordinate.
    pub fn transition(
        &mut self,
        target_s_coord: SCoord,
        new_gamma: Option<f64>,
        new_gamma_f: Option<f64>,
    ) -> TransitionResult {
        let from_state = self.current_state.s_coord;
        let distance = from_state
            .map(|s| s.distance(&target_s_coord))
            .unwrap_or(0.0);

        // Create new state
        let mut new_state = self.current_state.transition_to(target_s_coord);

        if let Some(g) = new_gamma {
            new_state = new_state.with_gamma(g);
        }
        if let Some(gf) = new_gamma_f {
            new_state = new_state.with_gamma_f(gf);
        }

        // Update history
        self.history.push(new_state.clone());
        if self.history.len() > self.max_history {
            self.history.remove(0);
        }

        self.current_state = new_state;

        TransitionResult {
            success: true,
            from_state,
            to_state: target_s_coord,
            distance,
            metadata: std::collections::HashMap::new(),
        }
    }

    /// Validate current state.
    pub fn validate_state(&self) -> bool {
        let state = &self.current_state;

        // Check bounds
        if state.gamma < 0.0 || state.gamma > 1.0 {
            return false;
        }
        if state.gamma_f < 0.0 || state.gamma_f > 1.0 {
            return false;
        }
        if state.p_decay < 0.0 || state.p_decay > 1.0 {
            return false;
        }
        if state.t_decay < 0.0 || state.t_decay > 1.0 {
            return false;
        }

        // Check S-coordinate bounds if present
        if let Some(s) = &state.s_coord {
            if s.sk < 0.0 || s.sk > 1.0 {
                return false;
            }
            if s.st < 0.0 || s.st > 1.0 {
                return false;
            }
            if s.se < 0.0 || s.se > 1.0 {
                return false;
            }
        }

        true
    }

    /// Extract S-coordinate trajectory.
    pub fn get_trajectory(&self) -> Vec<SCoord> {
        self.history
            .iter()
            .filter_map(|s| s.s_coord)
            .collect()
    }

    /// Extract consciousness time series.
    pub fn get_consciousness_history(&self) -> Vec<f64> {
        self.history.iter().map(|s| s.consciousness()).collect()
    }

    /// Extract memory time series.
    pub fn get_memory_history(&self) -> Vec<f64> {
        self.history.iter().map(|s| s.m).collect()
    }

    /// Find state at specific time.
    pub fn find_state_at_time(&self, t: f64) -> Option<&MentalState> {
        self.history.iter().find(|s| (s.timestamp - t).abs() < 1e-6)
    }

    /// Compute memory differential from history.
    pub fn compute_memory_differential(&self) -> f64 {
        if self.history.len() < 2 {
            return 0.0;
        }

        let n = self.history.len();
        let dt = self.history[n - 1].timestamp - self.history[n - 2].timestamp;
        if dt.abs() < 1e-12 {
            return 0.0;
        }

        (self.history[n - 1].m - self.history[n - 2].m) / dt
    }

    /// Enter dream mode.
    pub fn enter_dream_mode(&mut self) -> TransitionResult {
        let from_state = self.current_state.s_coord;
        self.current_state = self.current_state.enter_dream();

        self.history.push(self.current_state.clone());
        if self.history.len() > self.max_history {
            self.history.remove(0);
        }

        TransitionResult {
            success: true,
            from_state,
            to_state: self.current_state.s_coord.unwrap_or(SCoord::origin()),
            distance: 0.0,
            metadata: {
                let mut m = std::collections::HashMap::new();
                m.insert("mode".to_string(), "dream".to_string());
                m
            },
        }
    }

    /// Wake up from dream.
    pub fn wake_up(&mut self, perception_level: f64) -> TransitionResult {
        let from_state = self.current_state.s_coord;
        self.current_state = self.current_state.wake(perception_level);

        self.history.push(self.current_state.clone());
        if self.history.len() > self.max_history {
            self.history.remove(0);
        }

        TransitionResult {
            success: true,
            from_state,
            to_state: self.current_state.s_coord.unwrap_or(SCoord::origin()),
            distance: 0.0,
            metadata: {
                let mut m = std::collections::HashMap::new();
                m.insert("mode".to_string(), "awake".to_string());
                m
            },
        }
    }

    /// Clear history.
    pub fn clear_history(&mut self) {
        self.history.clear();
        self.history.push(self.current_state.clone());
    }

    /// Get summary statistics.
    pub fn summary(&self) -> std::collections::HashMap<String, f64> {
        let mut summary = std::collections::HashMap::new();

        summary.insert("current_consciousness".to_string(), self.current_state.consciousness());
        summary.insert("current_gamma".to_string(), self.current_state.gamma);
        summary.insert("current_gamma_f".to_string(), self.current_state.gamma_f);
        summary.insert("current_memory".to_string(), self.current_state.m);
        summary.insert("history_length".to_string(), self.history.len() as f64);

        if !self.history.is_empty() {
            let c_history = self.get_consciousness_history();
            let mean_c: f64 = c_history.iter().sum::<f64>() / c_history.len() as f64;
            summary.insert("mean_consciousness".to_string(), mean_c);
        }

        summary
    }
}

impl Default for StateManager {
    fn default() -> Self {
        Self::new(1000)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_initialize() {
        let mut manager = StateManager::new(100);
        let s = SCoord::new(0.5, 0.5, 0.5).unwrap();
        manager.initialize(Some(s), None);

        assert!(manager.current().s_coord.is_some());
        assert_eq!(manager.history().len(), 1);
    }

    #[test]
    fn test_transition() {
        let mut manager = StateManager::new(100);
        manager.initialize(Some(SCoord::origin()), None);

        let target = SCoord::new(0.8, 0.8, 0.8).unwrap();
        let result = manager.transition(target, Some(0.9), None);

        assert!(result.success);
        assert!(result.distance > 0.0);
        assert_eq!(manager.current().gamma, 0.9);
    }

    #[test]
    fn test_dream_wake() {
        let mut manager = StateManager::new(100);
        manager.initialize(Some(SCoord::origin()), None);

        manager.enter_dream_mode();
        assert!(manager.current().is_dreaming() || manager.current().p_decay < 0.1);

        manager.wake_up(0.9);
        assert!(manager.current().p_decay > 0.5);
    }

    #[test]
    fn test_history_limit() {
        let mut manager = StateManager::new(5);
        manager.initialize(Some(SCoord::origin()), None);

        for i in 0..10 {
            let s = SCoord::new(i as f64 / 10.0, 0.5, 0.5).unwrap();
            manager.transition(s, None, None);
        }

        assert!(manager.history().len() <= 6); // 5 + initial
    }
}
